site/content/blog/2020-12-04-common-i18n-mistakes.md

```diff
@@ -0,0 +1,315 @@
+
+```jsx
+return (
+    <>
+        <a
+            href='https://mattermost.com'
+            rel='noreferrer'
+            target='_blank'
+        >
+            <FormattedMessage
+                id='popup.clickHere'
+                defaultMessage='Click here'
+            />
+        </a>
+        {' '}
+        <FormattedMessage
+            id='popup.forMoreInformation'
+            defaultMessage='for more information.'
+        />
+    </>
+);
+```
+
+This works in English, but it may not work in other languages. For example, some languages may invert the sentence structure to be more like "For more information, click here" while others, like Chinese, may not include a space between words.
+
+Generally, we shouldn't be concatenating translated strings since it's a sign that the sentence structure may not work for everyone. Instead, we should move as much as we can into the translation string so that translators can construct the sentence as necessary. There's a few ways to do this that we've been iterating and improving on, so I'm going to present them separately and talk about why the final method is likely the best.
+
+### Solution 1: React elements as values
+
+Similar to the example on pluralization, we can actually pass translated text into another block of translated text as a value to construct the sentence.
+
+```jsx
+return (
+    <FormattedMessage
+        id='popup.moreInformation'
+        defaultMessage='{link} for more information.'
+        values={{
+            link: (
+                <a
+                    href='https://mattermost.com'
+                    rel='noreferrer'
+                    target='_blank'
+                >
+                    <FormattedMessage
+                        id='popup.moreInformation.link'
+                        defaultMessage='Click here'
+                    />
+                </a>
+            ),
+        }}
+    />
+);
+```
+
+You'll often see this used in older Mattermost code when we're trying to format part of a string. It allows us to add more complex formatting within translated text without making it impossible to translate.
+
+It does, however, make it harder for translators to understand the meaning of the text. They need to spend time reconstructing the entire sentence which may be difficult. It also makes the code more complicated to read for the same reason.
+
+### Solution 2: FormattedHTMLMessage
+
+`FormattedHTMLMessage` was another way to solve this problem, but it was removed in more recent versions. It's a compoent provided by react-intl, similar to `FormattedMessage`, but it allows us to include HTML in the translation string instead of nesting translated text. Since it's been removed, it can't be used any more, but it demonstrates another problem with ran into in the past.
```

Comment:
```suggestion
`FormattedHTMLMessage` was another way to solve this problem, but it was removed in more recent versions. It's a component provided by react-intl, similar to `FormattedMessage`, but it allows us to include HTML in the translation string instead of nesting translated text. Since it's been removed, it can't be used any more, but it demonstrates another problem with ran into in the past.
```